modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/model/listener/CPInstanceUnitOfMeasureModelListener.java

```diff
@@ -0,0 +1,102 @@
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.commerce.internal.model.listener;
+
+import com.liferay.commerce.constants.CommerceOrderConstants;
+import com.liferay.commerce.model.CommerceOrderItem;
+import com.liferay.commerce.product.model.CPInstanceUnitOfMeasure;
+import com.liferay.commerce.product.service.CPInstanceUnitOfMeasureLocalService;
+import com.liferay.commerce.service.CommerceOrderItemLocalService;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.dao.orm.QueryUtil;
+import com.liferay.portal.kernel.exception.ModelListenerException;
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.util.ArrayUtil;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Stefano Motta
+ */
+@Component(service = ModelListener.class)
+public class CPInstanceUnitOfMeasureModelListener
+	extends BaseModelListener<CPInstanceUnitOfMeasure> {
+
+	@Override
+	public void onAfterCreate(CPInstanceUnitOfMeasure cpInstanceUnitOfMeasure) {
+		int cpInstanceUnitOfMeasuresCount =
+			_cpInstanceUnitOfMeasureLocalService.
+				getCPInstanceUnitOfMeasuresCount(
+					cpInstanceUnitOfMeasure.getCPInstanceId());
+
+		if (cpInstanceUnitOfMeasuresCount != 1) {
+			return;
+		}
+
+		_updateUnitOfMeasureKey(
+			cpInstanceUnitOfMeasure.getCPInstanceId(),
+			cpInstanceUnitOfMeasure.getKey(), StringPool.BLANK);
+	}
+
+	@Override
+	public void onAfterUpdate(
+			CPInstanceUnitOfMeasure originalCPInstanceUnitOfMeasure,
+			CPInstanceUnitOfMeasure cpInstanceUnitOfMeasure)
+		throws ModelListenerException {
+
+		String key = cpInstanceUnitOfMeasure.getKey();
+
+		if (key.equals(originalCPInstanceUnitOfMeasure.getKey())) {
+			return;
+		}
+
+		int cpInstanceUnitOfMeasuresCount =
+			_cpInstanceUnitOfMeasureLocalService.
+				getCPInstanceUnitOfMeasuresCount(
+					cpInstanceUnitOfMeasure.getCPInstanceId());
+
+		if (cpInstanceUnitOfMeasuresCount != 1) {
+			return;
+		}
+
+		_updateUnitOfMeasureKey(
+			originalCPInstanceUnitOfMeasure.getCPInstanceId(),
+			cpInstanceUnitOfMeasure.getKey(),
+			originalCPInstanceUnitOfMeasure.getKey());
+	}
+
+	private void _updateUnitOfMeasureKey(
+		long cpInstanceId, String key, String originalKey) {
+
+		int[] orderStatuses = ArrayUtil.append(
+			CommerceOrderConstants.ORDER_STATUSES_OPEN,
+			CommerceOrderConstants.ORDER_STATUSES_PENDING,
+			CommerceOrderConstants.ORDER_STATUSES_PROCESSING,
+			CommerceOrderConstants.ORDER_STATUSES_SHIPPING);
+
+		for (CommerceOrderItem commerceOrderItem :
+				_commerceOrderItemLocalService.
+					getCommerceOrderItemsForCommerceOrdersWithStatuses(
+						cpInstanceId, orderStatuses, originalKey,
+						QueryUtil.ALL_POS, QueryUtil.ALL_POS)) {
+
+			commerceOrderItem.setUnitOfMeasureKey(key);
```

Comment:
what about the quantity?

Reply:
Now we multiply the quantity with the incrementalOrderQuantity